python/ray/experimental/data/datasource.py

```diff
@@ -166,6 +265,108 @@ def make_block(start: int, count: int) -> Block[Union[ArrowRow, int]]:
+
+        def read_json(read_paths: List[str]):
+            logger.debug(f"Reading {len(read_paths)} files.")
+            tables = []
+            for read_path in read_paths:
+                with filesystem.open_input_file(read_path) as f:
+                    tables.append(
+                        json.read_json(
+                            f,
+                            read_options=json.ReadOptions(use_threads=False),
+                            **arrow_json_args))
+            block = ArrowBlock(pa.concat_tables(tables))
+            return block
+
+        read_tasks = [
+            ReadTask(
+                lambda read_paths=read_paths: read_json(read_paths),
+                BlockMetadata(
+                    num_rows=None,
+                    size_bytes=sum(file_sizes),
+                    schema=None,
+                    input_files=read_paths)) for read_paths, file_sizes in zip(
+                        np.array_split(paths, parallelism),
+                        np.array_split(file_sizes, parallelism))
+            if len(read_paths) > 0
+        ]
+
+        return read_tasks
+
+
+class CSVDatasource(Datasource[Union[ArrowRow, int]]):
+    """CSV datasource, for reading CSV files.
+
+    Examples:
+        >>> source = CSVDatasource()
+        >>> ray.data.read_datasource(source, paths="/path/to/dir").take()
+        ... {"a": 1, "b": "foo"}
+    """
+
+    def prepare_read(self,
+                     parallelism: int,
+                     paths: Union[str, List[str]],
+                     filesystem: Optional["pyarrow.fs.FileSystem"] = None,
+                     **arrow_csv_args) -> List[ReadTask]:
+        import pyarrow as pa
+        from pyarrow import csv
+        import numpy as np
+
+        paths, file_infos, filesystem = _resolve_paths_and_filesystem(
+            paths, filesystem)
+        file_sizes = [file_info.size for file_info in file_infos]
+
+        def read_csv(read_paths: List[str]):
+            logger.debug(f"Reading {len(read_paths)} files.")
+            tables = []
+            for read_path in read_paths:
+                with filesystem.open_input_file(read_path) as f:
+                    tables.append(
+                        csv.read_csv(
+                            f,
+                            read_options=csv.ReadOptions(use_threads=False),
+                            **arrow_csv_args))
+            block = ArrowBlock(pa.concat_tables(tables))
+            return block
+
+        read_tasks = [
+            ReadTask(
+                lambda read_paths=read_paths: read_csv(read_paths),
+                BlockMetadata(
+                    num_rows=None,
+                    size_bytes=sum(file_sizes),
+                    schema=None,
+                    input_files=read_paths)) for read_paths, file_sizes in zip(
+                        np.array_split(paths, parallelism),
+                        np.array_split(file_sizes, parallelism))
+            if len(read_paths) > 0
```

Comment:
Can we make a `FileBasedDatasource` common class in `file_based_datasource.py` that both csv / json can use as a base class?